dali/kernels/common/simd.h

```diff
@@ -0,0 +1,209 @@
+  __m128i i32_1 = _mm_unpackhi_epi8(lo16, zero);
+  __m128i i32_2 = _mm_unpacklo_epi8(hi16, zero);
+  __m128i i32_3 = _mm_unpackhi_epi8(hi16, zero);
+  return {{ _mm_cvtepi32_ps(i32_0),
+            _mm_cvtepi32_ps(i32_1),
+            _mm_cvtepi32_ps(i32_2),
+            _mm_cvtepi32_ps(i32_3) }};
+}
+
+inline __m128i sext8_32(__m128i i) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 24);
+  return _mm_sra_epi32(_mm_sll_epi32(i, sh), sh);
+}
+
+inline float4x4 load_f(const int8_t *i8) {
+  __m128i in = _mm_loadu_si128(reinterpret_cast<const __m128i *>(i8));
+  __m128i zero = _mm_setzero_si128();
+  __m128i lo16 = _mm_unpacklo_epi8(in, zero);
+  __m128i hi16 = _mm_unpackhi_epi8(in, zero);
+  __m128i i32_0 = _mm_unpacklo_epi8(lo16, zero);
+  __m128i i32_1 = _mm_unpackhi_epi8(lo16, zero);
+  __m128i i32_2 = _mm_unpacklo_epi8(hi16, zero);
+  __m128i i32_3 = _mm_unpackhi_epi8(hi16, zero);
+  return {{ _mm_cvtepi32_ps(sext8_32(i32_0)),
+            _mm_cvtepi32_ps(sext8_32(i32_1)),
+            _mm_cvtepi32_ps(sext8_32(i32_2)),
+            _mm_cvtepi32_ps(sext8_32(i32_3)) }};
+}
+
+inline __m128i sext16_32(__m128i i) {
+  __m128i sh = _mm_set_epi32(0, 0, 0, 16);
+  return _mm_sra_epi32(_mm_sll_epi32(i, sh), sh);
+}
+
+inline float4x2 load_f(const uint16_t *u16) {
+  __m128i in = _mm_loadu_si128(reinterpret_cast<const __m128i *>(u16));
+  __m128i zero = _mm_setzero_si128();
+  __m128i i32_0 = _mm_unpacklo_epi16(in, zero);
+  __m128i i32_1 = _mm_unpackhi_epi16(in, zero);
+  return {{ _mm_cvtepi32_ps(i32_0), _mm_cvtepi32_ps(i32_1) }};
+}
+
+inline float4x2 load_f(const int16_t *i16) {
+  __m128i in = _mm_loadu_si128(reinterpret_cast<const __m128i *>(i16));
+  __m128i zero = _mm_setzero_si128();
+  __m128i i32_0 = _mm_unpacklo_epi16(in, zero);
+  __m128i i32_1 = _mm_unpackhi_epi16(in, zero);
+  return {{ _mm_cvtepi32_ps(sext16_32(i32_0)), _mm_cvtepi32_ps(sext16_32(i32_1)) }};
+}
+
+inline float4x1 load_f(const float *f) {
+  return {{ _mm_loadu_ps(f) }};
+}
+
+template <typename Out>
+inline std::enable_if_t<std::is_integral<Out>::value>
+store(Out *out, float4x<sizeof(float)/sizeof(Out)> f) {
+  constexpr int nvec = sizeof(float)/sizeof(Out);
+  constexpr float lo = min_value<Out>();
+  constexpr float hi = max_value<Out>();
+  i128x<nvec> iv;
+  for (int i = 0; i < nvec; i++)
+    iv.v[i] = clamp_round(f.v[i], lo, hi);
+  store(out, iv);
+}
+
+inline void store(float *out, float4x1 f) {
+  _mm_storeu_ps(out, f.v[0]);
+}
+
+template <int num_vecs>
+struct multivec : float4x<num_vecs> {
+  DALI_FORCEINLINE static multivec zero() noexcept  {
+    multivec m;
+    for (int i = 0; i < num_vecs; i++)
+      m.v[i] = _mm_setzero_ps();
+    return m;
+  }
+
+  DALI_FORCEINLINE static multivec load(const float *in) noexcept  {
+    multivec m;
+    for (int i = 0; i < num_vecs; i++)
+      m.v[i] = _mm_loadu_ps(in + 4*i);
+    return m;
+  }
+
+  template <typename In>
+  DALI_FORCEINLINE static multivec load(const In *in) noexcept  {
```

Comment:
Don't you have to specify the correct `num_vecs` based on the input type?
Can't you compute it based on the input type instead? It seems error prone to me.

Reply:
No. It will load as many values from input as is necessary to fill the multivec.
I can add
```C++
static_assert(load_vecs > 0, "Too few lanes to use vector load for this input type");
```

Reply:
I mean, if you try to load uint16 into the float4x1, it will just probably segfault. There is no check for that, IMO it should be
```
static_assert(num_vecs % load_vecs == 0);
```
or something similar.

Reply:
Not the implementation of multivec is tied to it's usage that assumes the correct storage size used for loading.

Reply:
Fixed?